### PR TITLE
feat(es‑abstract): Make `ESAbstract.PropertyDescriptor` generic

### PR DESCRIPTION
**Pull request by ExE-Boss**
This makes the `ESAbstract.PropertyDescriptor` definition generic to improve type safety and match `TypedPropertyDescriptor`.

## Blocks:
- <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44805>

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes:
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/es-abstract/es2015.d.ts [hunk 1]
```diff
@@ -225,9 +250,9 @@ declare namespace ES2015 {
     // Re-export types from previous versions
     // - ES5:
     type GenericDescriptor = ES5.GenericDescriptor;
-    type AccessorDescriptor = ES5.AccessorDescriptor;
-    type DataDescriptor = ES5.DataDescriptor;
-    type PropertyDescriptor = ES5.PropertyDescriptor;
+    type AccessorDescriptor<T = any> = ES5.AccessorDescriptor<T>;
+    type DataDescriptor<T = any> = ES5.DataDescriptor<T>;
+    type PropertyDescriptor<T = any> = ES5.PropertyDescriptor<T>;
```
Comment (ljharb) [inline types/es-abstract/es2015.d.ts:255]:
```suggestion
    type AccessorDescriptor<T = unknown> = ES5.AccessorDescriptor<T>;
    type DataDescriptor<T = unknown> = ES5.DataDescriptor<T>;
    type PropertyDescriptor<T = unknown> = ES5.PropertyDescriptor<T>;
```

File: types/es-abstract/es2015.d.ts [hunk 2]
```diff
@@ -191,7 +200,23 @@ interface ES2015 extends Omit<typeof ES5, 'CheckObjectCoercible' | 'ToPrimitive'
...
+            '[[Configurable]]': boolean;
+            '[[Enumerable]]': boolean;
+            '[[Writable]]': boolean;
+            '[[Value]]': P extends keyof O ? O[P] : any;
+        }
+        | {
+            '[[Configurable]]': boolean;
+            '[[Enumerable]]': boolean;
+            '[[Get]]': (() => P extends keyof O ? O[P] : any) | undefined;
+            '[[Set]]': (value: P extends keyof O ? O[P] : any) => void;
```
Comment (ljharb) [inline types/es-abstract/es2015.d.ts:217]:
```suggestion
            '[[Get]]': (() => P extends keyof O ? O[P] : unknown) | undefined;
            '[[Set]]': (value: P extends keyof O ? O[P] : unknown) => void;
```

File: types/es-abstract/es2015.d.ts [hunk 3]
```diff
@@ -191,7 +200,23 @@ interface ES2015 extends Omit<typeof ES5, 'CheckObjectCoercible' | 'ToPrimitive'
...
-    OrdinaryGetOwnProperty(O: object, P: ESPropertyKey): ES5.PropertyDescriptor | undefined;
+    OrdinaryGetOwnProperty<O extends object, P extends ESPropertyKey>(
+        O: O,
+        P: P,
+    ):
+        | {
+            '[[Configurable]]': boolean;
+            '[[Enumerable]]': boolean;
+            '[[Writable]]': boolean;
+            '[[Value]]': P extends keyof O ? O[P] : any;
```
Comment (ljharb) [inline types/es-abstract/es2015.d.ts:211]:
```suggestion
            '[[Value]]': P extends keyof O ? O[P] : unknown;
```

**Review by ljharb — COMMENTED** [review]
Same theme re replacing all any with unknown, otherwise LGTM

File: types/es-abstract/es5.d.ts [hunk 4]
```diff
@@ -79,9 +79,9 @@ interface ES5 {
 
 declare namespace ES5 {
     type GenericDescriptor = ESGenericDescriptor;
-    type AccessorDescriptor = ESAccessorDescriptor;
-    type DataDescriptor = ESDataDescriptor;
-    type PropertyDescriptor = ESPropertyDescriptor;
+    type AccessorDescriptor<T = any> = ESAccessorDescriptor<T>;
+    type DataDescriptor<T = any> = ESDataDescriptor<T>;
+    type PropertyDescriptor<T = any> = ESPropertyDescriptor<T>;
```
Comment (ljharb) [inline types/es-abstract/es5.d.ts:84]:
```suggestion
    type AccessorDescriptor<T = unknown> = ESAccessorDescriptor<T>;
    type DataDescriptor<T = unknown> = ESDataDescriptor<T>;
    type PropertyDescriptor<T = unknown> = ESPropertyDescriptor<T>;
```

**Review by ljharb — APPROVED** [review]
One more :-)

File: types/es-abstract/es2015.d.ts [hunk 5]
```diff
@@ -71,10 +69,21 @@ interface ES2015 extends Omit<typeof ES5, 'CheckObjectCoercible' | 'ToPrimitive'
     CompletePropertyDescriptor<D extends ES5.PropertyDescriptor>(
         Desc: D & ThisType<any>,
```
Comment (ljharb) [inline types/es-abstract/es2015.d.ts:70]: i'd prefer all these any's changed to unknown's also, but that can be in a separate PR too

**Review by ljharb — APPROVED** [review]

**Comment by ExE-Boss** [comment]
Ready to merge

File: types/es-abstract/es2015.d.ts (hunk 5 above)
Comment (ExE-Boss) [reply types/es-abstract/es2015.d.ts:70]: I intend to do it in <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44805>.
